src/physics/em/detail/RayleighInteractor.i.hh

```diff
@@ -15,54 +18,123 @@ namespace detail
-    , allocate_(allocate)
+    , inc_direction_(direction)
+    , element_id_(el_id)
 {
     CELER_EXPECT(inc_energy_ >= this->min_incident_energy()
                  && inc_energy_ <= this->max_incident_energy());
-    CELER_EXPECT(particle.particle_id() == shared_.gamma_id); // XXX
-    CELER_NOT_IMPLEMENTED("Rayleigh scattering");
+    CELER_EXPECT(particle.particle_id() == shared_.gamma_id);
 }
 
 //---------------------------------------------------------------------------//
 /*!
- * Sample using the XXX model.
+ * Sample the Rayleigh scattering angle using the G4LivermoreRayleighModel
+ * and G4RayleighAngularGenerator of Geant4 6.10.
  */
 template<class Engine>
 CELER_FUNCTION Interaction RayleighInteractor::operator()(Engine& rng)
 {
-    // Allocate space for XXX (electron, multiple particles, ...)
-    Secondary* secondaries = this->allocate_(0); // XXX
-    if (secondaries == nullptr)
-    {
-        // Failed to allocate space for a secondary
-        return Interaction::from_failure();
-    }
-
-    // XXX sample
-    (void)sizeof(rng);
+    real_type energy = inc_energy_.value();
 
     // Construct interaction for change to primary (incident) particle
     Interaction result;
-    result.action      = Action::scattered;                     // XXX
-    result.energy      = units::MevEnergy{inc_energy_.value()}; // XXX
-    result.direction   = inc_direction_;
-    result.secondaries = {secondaries, 1}; // XXX
+    result.action = Action::scattered;
+    result.energy = units::MevEnergy{inc_energy_.value()};
+
+    // Sample direction for a given Z: G4RayleighAngularGenerator
+    ItemIdT item_id = ItemIdT{element_id_.get()};
+
+    Real3 pn = shared_.params.data_n[item_id];
+    Real3 pb = shared_.params.data_b[item_id];
+    Real3 px = shared_.params.data_x[item_id];
+
+    SampleInput input = this->evaluate_weight_and_prob(energy, pb, pn, px);
+
+    real_type                          cost;
+    UniformRealDistribution<real_type> u01(0, 1.0);
+
+    do
+    {
+        unsigned  index = 0;
+        real_type x     = u01(rng);
+
+        if (x > input.prob[0])
+        {
+            index = (x <= input.prob[0] + input.prob[1]) ? 1 : 2;
+        }
+
+        real_type w = input.weight[index];
+        real_type n = pn[index];
+        real_type b = pb[index];
+
+        n = 1.0 / n;
+
+        // Sampling of scattering angle
+        real_type y = w * u01(rng);
+        if (y < num_limit())
+        {
+            x = y * n * (1. + 0.5 * (n + 1.) * y * (1. - (n + 2.) * y / 3.));
+        }
+        else
+        {
+            x = std::exp(-n * std::log(1. - y)) - 1.0;
+        }
 
-    // Save outgoing secondary data
-    secondaries[0].particle_id = shared_.electron_id; // XXX
-    secondaries[0].energy      = units::MevEnergy{0}; // XXX
-    secondaries[0].direction   = {0, 0, 0};           // XXX
+        cost = 1.0 - 2.0 * x / (b * input.factor);
+
+    } while (2 * u01(rng) > 1.0 + cost * cost || cost < -1.0);
```

Comment:
Use `ipow<2>(cost)` for the inline "squared". And is `cost` ever really going to be less than -1? I think some extra sanity assertions (bounds checking) in here would be good.

Reply:
Used ipow<2>(cost).  Regarding to cost, it seems that there are cases that cost < -1.0 during the sampling (so cannot asset it during the sampling).  I will leave this as it is for now.